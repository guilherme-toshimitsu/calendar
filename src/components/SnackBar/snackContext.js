import { createContext } from 'react';

const SnackBarContext = createContext({
  snackValues: {
    vertical: 'top',
    horizontal: 'right',
    snackbarType: 'success',
    message: 'general.ok',
    autoClose: false,
    onSuccess: () => {},
  },
  openSnackBar: () => {},
});
const SnackBarProvider = SnackBarContext.Provider;
const SnackBarConsumer = SnackBarContext.Consumer;

export default SnackBarContext;
export { SnackBarConsumer, SnackBarProvider, SnackBarContext };
