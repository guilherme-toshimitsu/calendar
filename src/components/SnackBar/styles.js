import green from '@material-ui/core/colors/green';
import amber from '@material-ui/core/colors/amber';
import blue from '@material-ui/core/colors/blue';

const snackBarContentStyles = (theme) => ({
  success: {
    backgroundColor: green[600],
  },
  error: {
    backgroundColor: theme.palette.error.dark,
  },
  info: {
    backgroundColor: blue[600],
  },
  warning: {
    backgroundColor: amber[700],
  },
  icon: {
    fontSize: 20,
  },
  iconVariant: {
    opacity: 0.9,
    marginRight: theme.spacing(),
  },
  message: {
    display: 'flex',
    alignItems: 'center',
    alignText: 'left',
  },
});

const snackBarStyles = (theme) => ({
  root: {
    maxWidth: '70%',
  },
  close: {
    padding: theme.spacing() / 2,
  },
});

export { snackBarStyles, snackBarContentStyles };
