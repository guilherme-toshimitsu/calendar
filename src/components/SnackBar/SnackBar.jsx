import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

import {
  IconButton,
  SnackbarContent,
  Snackbar as MuiSnackbar,
  makeStyles,
} from '@material-ui/core';

import {
  Close as CloseIcon,
  Warning as WarningIcon,
  Error as ErrorIcon,
  Info as InfoIcon,
  CheckCircle as CheckCircleIcon,
} from '@material-ui/icons';

import { snackBarContentStyles, snackBarStyles } from './styles';

const variantIcon = {
  success: CheckCircleIcon,
  warning: WarningIcon,
  error: ErrorIcon,
  info: InfoIcon,
};

const useStylesSnackbarContent = makeStyles(snackBarContentStyles, {
  name: 'snackBarContent',
});
const useStylesSnackbar = makeStyles(snackBarStyles, { name: 'snackBar' });

// eslint-disable-next-line react/display-name
const MySnackbarContent = forwardRef(
  (
    { className, message, onClose, variant, onSuccess, action, ...other },
    ref
  ) => {
    const Icon = variantIcon[variant]
      ? variantIcon[variant]
      : variantIcon.error;
    const classes = useStylesSnackbarContent();

    const actionComponent = (action) =>
      action || [
        <IconButton
          key="close"
          aria-label="Close"
          color="inherit"
          className={classes.close}
          onClick={variant === 'error' ? onClose : onSuccess}
          data-testid="snackbar-icon-button-close"
        >
          <CloseIcon className={classes.icon} />
        </IconButton>,
      ];

    return (
      <SnackbarContent
        ref={ref}
        className={clsx(classes[variant], className)}
        aria-describedby="client-snackbar"
        message={
          <span id="client-snackbar" className={classes.message}>
            <Icon className={clsx(classes.icon, classes.iconVariant)} />
            <span dangerouslySetInnerHTML={{ __html: message }} />
          </span>
        }
        action={actionComponent(action)}
        {...other}
      />
    );
  }
);

MySnackbarContent.defaultProps = {
  action: null,
};

MySnackbarContent.propTypes = {
  className: PropTypes.string,
  message: PropTypes.string,
  onClose: PropTypes.func,
  action: PropTypes.any,
  variant: PropTypes.oneOf(['success', 'warning', 'error', 'info']).isRequired,
  onSuccess: PropTypes.func.isRequired,
};

const SnackBar = ({
  vertical,
  horizontal,
  open,
  action,
  message,
  onClose,
  snackbarType,
  onSuccess,
  ...props
}) => {
  const classes = useStylesSnackbar();
  return (
    <MuiSnackbar
      classes={Object.assign({}, { root: classes.root }, props.classes)}
      anchorOrigin={{ vertical, horizontal }}
      open={open}
      onClose={onClose}
      ContentProps={{
        'aria-describedby': 'message-id',
      }}
    >
      <MySnackbarContent
        onClose={onClose}
        variant={snackbarType}
        message={message}
        onSuccess={onSuccess}
        action={action}
      />
    </MuiSnackbar>
  );
};

SnackBar.propTypes = {
  vertical: PropTypes.string,
  horizontal: PropTypes.string,
  open: PropTypes.bool.isRequired,
  message: PropTypes.string,
  action: PropTypes.any,
  onClose: PropTypes.func.isRequired,
  snackbarType: PropTypes.oneOf(['success', 'error', 'warning', 'info']),
  onSuccess: PropTypes.func,
  classes: PropTypes.object,
};

SnackBar.defaultProps = {
  classes: {},
  vertical: 'top',
  horizontal: 'right',
  snackbarType: 'success',
  message: '',
  action: null,
  onSuccess: () => {},
};

export default SnackBar;
export { MySnackbarContent };
