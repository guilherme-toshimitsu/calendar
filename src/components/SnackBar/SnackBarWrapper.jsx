import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import SnackBar from './SnackBar';
import { SnackBarProvider } from './snackContext';

const initialValuesSnackBar = {
  vertical: 'top',
  horizontal: 'right',
  snackbarType: 'success',
  message: 'general.ok',
  autoClose: false,
  action: null,
  onSuccess: () => {},
};

const SnackBarWrapper = ({ children }) => {
  const [snackValues, setSnackValues] = useState(initialValuesSnackBar);
  const [open, setOpen] = useState(false);

  const openSnackBar = (val) => {
    const snack = Object.assign(initialValuesSnackBar, val);
    setSnackValues(snack);
    setOpen(true);
  };

  const onSuccess = () => {
    snackValues.onSuccess();
    setOpen(false);
  };

  useEffect(() => {
    if (open === true && snackValues.autoClose) {
      setTimeout(() => {
        setOpen(false);
        if (snackValues.snackbarType === 'success') {
          snackValues.onSuccess();
        }
      }, snackValues.autoClose);
    }
  }, [open, snackValues.autoClose, setOpen, snackValues]);

  return (
    <SnackBarProvider
      value={{
        snackValues: snackValues,
        openSnackBar,
      }}
    >
      <SnackBar
        open={open}
        onClose={() => setOpen(false)}
        onSuccess={() => onSuccess()}
        vertical={snackValues.vertical}
        horizontal={snackValues.horizontal}
        snackbarType={snackValues.snackbarType}
        action={snackValues.action}
        message={snackValues.message}
      />
      {children}
    </SnackBarProvider>
  );
};

SnackBarWrapper.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

export default SnackBarWrapper;
