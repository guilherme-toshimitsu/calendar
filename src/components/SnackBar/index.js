import { useContext } from 'react';
import SnackBar from './SnackBar';
import SnackWrapper from './SnackBarWrapper';
import { SnackBarContext } from './snackContext';

const useSnackBar = () => useContext(SnackBarContext);

export default SnackBar;
export { useSnackBar, SnackWrapper };
