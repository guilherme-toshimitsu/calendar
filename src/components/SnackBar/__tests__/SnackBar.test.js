import React from 'react';
import { render } from '@testing-library/react';
import { createMuiTheme } from '@material-ui/core';

import SnackBar, { MySnackbarContent } from '../SnackBar';

const theme = {};

describe('<SnackBar />', () => {
  test('if render', () => {
    const { asFragment } = render(
      <SnackBar
        vertical="top"
        horizontal="center"
        open={false}
        action={() => {}}
        message="msg"
        onClose={() => {}}
        snackbarType="success"
        onSuccess={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<MySnackBarContent />', () => {
  test('if render', () => {
    const { asFragment } = render(
      <MySnackbarContent
        vertical="top"
        horizontal="center"
        open={false}
        action={() => {}}
        message="msg"
        onClose={() => {}}
        variant="success"
        onSuccess={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('if render without action', () => {
    const { asFragment } = render(
      <MySnackbarContent
        vertical="top"
        horizontal="center"
        open={false}
        message="msg"
        onClose={() => {}}
        onSuccess={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
