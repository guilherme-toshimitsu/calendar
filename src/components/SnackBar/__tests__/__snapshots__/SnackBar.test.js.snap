// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MySnackBarContent /> if render 1`] = `
<DocumentFragment>
  <div
    aria-describedby="client-snackbar"
    class="MuiPaper-root MuiSnackbarContent-root snackBarContent-success-10 MuiPaper-elevation6"
    horizontal="center"
    role="alert"
    vertical="top"
  >
    <div
      class="MuiSnackbarContent-message"
    >
      <span
        class="snackBarContent-message-16"
        id="client-snackbar"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root snackBarContent-icon-14 snackBarContent-iconVariant-15"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span>
          msg
        </span>
      </span>
    </div>
    <div
      class="MuiSnackbarContent-action"
    />
  </div>
</DocumentFragment>
`;

exports[`<MySnackBarContent /> if render without action 1`] = `
<DocumentFragment>
  <div
    aria-describedby="client-snackbar"
    class="MuiPaper-root MuiSnackbarContent-root MuiPaper-elevation6"
    horizontal="center"
    role="alert"
    vertical="top"
  >
    <div
      class="MuiSnackbarContent-message"
    >
      <span
        class="snackBarContent-message-63"
        id="client-snackbar"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root snackBarContent-icon-61 snackBarContent-iconVariant-62"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
          />
        </svg>
        <span>
          msg
        </span>
      </span>
    </div>
    <div
      class="MuiSnackbarContent-action"
    >
      <button
        aria-label="Close"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        data-testid="snackbar-icon-button-close"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root snackBarContent-icon-61"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SnackBar /> if render 1`] = `<DocumentFragment />`;
